sdk/ml/azure-ai-ml/azure/ai/ml/entities/_builders/base_node.py

```diff
@@ -303,29 +288,30 @@ def _get_component_id(self) -> Union[str, Component]:
             # If component is remote, return it's asset id
             return self._component.id
         # Otherwise, return the component version or arm id.
-        return self._component
+        res: Union[str, Component] = self._component
+        return res
 
-    def _get_component_name(self):
+    def _get_component_name(self) -> Optional[str]:
         # first use component version/job's display name or name as component name
         # make it unique when pipeline build finished.
         if self._component is None:
             return None
         if isinstance(self._component, str):
             return self._component
-        return self._component.name
+        return str(self._component.name)
 
     def _to_dict(self) -> Dict:
-        return convert_ordered_dict_to_dict(self._dump_for_validation())
+        return dict(convert_ordered_dict_to_dict(self._dump_for_validation()))
```

Comment:
We should not use dict to force conversion to dict. Instead, we should ensure that the return value is dict by using `convertreordered_dit_to_dict` function.

Reply:
@CoderKevinZhang - perhaps the better solution here is to ensure that the return type of `convert_ordered_dict_to_dict` is annotated as a `Dict`. Then, we wouldn't need the cast.

Reply:
Annotate `convert_ordered_dict_to_dict` and also type this return if that's still not enough

```python
converted_dict: Dict = convert_ordered_dict_to_dict(self._dump_for_validation())
return converted_dict
```

Reply:
Fixed